source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -336,10 +336,39 @@ namespace
         return elem;
     }
 
-    void readDocument(const DocumentPtr& doc, PrvStage* stage, RtFileIo::ReadFilter filter)
+    void readSourceUri(const DocumentPtr& doc, PrvStage* stage)
     {
+        StringSet uris = doc->getReferencedSourceUris();
+        RtTokenList& stageUris = stage->getSourceUri();
+        for (const string& uri : uris)
+        {
+            stageUris.push_back(RtToken(uri));
+        }
+    }
+
+    void readDocument(const DocumentPtr& doc, PrvStage* stage, const RtReadOptions* readOptions)
+    {
+        // Set the source location 
+        const std::string& uri = doc->getSourceUri();
+        RtTokenList uris = stage->getSourceUri();
+        uris.push_back(RtToken(uri));
         readAttributes(doc, stage, {});
 
+        RtReadOptions::ReadFilter filter = readOptions ? readOptions->readFilter : nullptr;
+
+        // First, load all nodedefs. Having these available is needed
+        // when node instances are loaded later.
+        for (const NodeDefPtr& nodedef : doc->getNodeDefs())
```

Comment:
If there are nodedefs here, then we need to do similar logic as for loadlibraries() to preload the nodedefs.